src/emplacer.rs

```diff
@@ -0,0 +1,288 @@
+        let invoice_path = self.invoice_path(id);
+        if !invoice_path.is_file() {
+            let invoice_text = reader.get_invoice_bytes(id).await?;
+            safely_write(&invoice_path, invoice_text).await
+                .with_context(|| format!("Error writing invoice {} to cache", &id))?;
+        }
+
+        let invoice_text = tokio::fs::read(&invoice_path).await
+            .with_context(|| format!("Error reading cached invoice file {}", invoice_path.display()))?;
+        let invoice = toml::from_slice(&invoice_text)
+            .with_context(|| format!("Error parsing cached invoice file {}", invoice_path.display()))?;
+
+        let invoice = InvoiceUnderstander::new(&invoice);
+
+        let module_parcels: Vec<_> = invoice
+            .top_modules().iter()
+            .filter_map(|parcel| invoice.classify_parcel(parcel))
+            .filter_map(|parcel| match parcel {
+                InterestingParcel::WagiHandler(h) => Some(h),
+            })
+            .collect();
+
+        let module_placements = module_parcels.iter().map(|h| self.emplace_module_and_assets(reader, &id, &h));
+        let all_module_placements = futures::future::join_all(module_placements).await;
+
+        match all_module_placements.into_iter().find_map(|e| e.err()) {
+            Some(e) => Err(e),
+            None => Ok(())
+        }
+    }
+
+    async fn emplace_module_and_assets(&self, reader: &impl BindleReader, invoice_id: &bindle::Id, handler: &WagiHandlerInfo) -> anyhow::Result<()> {
+        self.emplace_module(reader, invoice_id, &handler.parcel).await?;
+        self.emplace_as_assets(reader, invoice_id, &handler.asset_parcels()).await?;
+        Ok(())
+    }
+
+    async fn emplace_module(&self, reader: &impl BindleReader, invoice_id: &bindle::Id, parcel: &bindle::Parcel) -> anyhow::Result<()> {
+        let parcel_path = self.cache_path.join(&parcel.label.sha256);
+        if parcel_path.is_file() {
+            return Ok(());
+        }
+
+        let parcel_data = reader.get_parcel(invoice_id, &parcel).await?;
+        safely_write(&parcel_path, parcel_data).await
+            .with_context(|| format!("Error caching parcel {} at {}", parcel.label.name, parcel_path.display()))
+    }
+
+    async fn emplace_as_asset(&self, reader: &impl BindleReader, invoice_id: &bindle::Id, parcel: &bindle::Parcel) -> anyhow::Result<()> {
+        let parcel_path = self.asset_parcel_path(invoice_id, parcel);
+        if parcel_path.is_file() {
+            return Ok(());
+        }
+
+        let parcel_data = reader.get_parcel(invoice_id, &parcel).await?;
+        safely_write(&parcel_path, parcel_data).await
+            .with_context(|| format!("Error caching parcel {} at {}", parcel.label.name, parcel_path.display()))?;
+        Ok(())
+    }
+
+    async fn emplace_as_assets(&self, reader: &impl BindleReader, invoice_id: &bindle::Id, parcels: &Vec<bindle::Parcel>) -> anyhow::Result<()> {
```

Comment:
Given the struct is already an "Emplacer", what do you think about simplifying the names of all the methods that also contain "emplace_"?

Reply:
Open to suggestions but just removing the `emplace` prefix would reduce clarity.